File: C#/exercism/ProteinTranslation.cs
namespace exercism;

public static class ProteinTranslation
{
    private static string Mapping(string? s)
        => s switch
        {
            "AUG" => "Methionine",
            "UUU" or "UUC" => "Phenylalanine",
            "UUA" or "UUG" => "Leucine",
            "UCU" or "UCC" or "UCA" or "UCG" => "Serine",
            "UAU" or "UAC" => "Tyrosine",
            "UGU" or "UGC" => "Cysteine",
            "UGG" => "Tryptophan",
            "UAA" or "UAG" or "UGA" => "STOP",
            _ => ""
        };
    public static string[] Proteins(string strand) =>
        strand
            .Chunk(3)
            .Select(entry => string.Concat(entry))
            .Select(Mapping)
            .TakeWhile(entry => entry != "STOP").ToArray();
}

File: F#/Exercism/proteinTranslation.fs
module proteinTranslation


let translate (code: string): string = 
    match code with
        | "AUG" -> "Methionine"
        | "UUU" | "UUC" -> "Phenylalanine"
        | "UUA" | "UUG" -> "Leucine"
        | "UCU" | "UCC" | "UCA" | "UCG" -> "Serine"
        | "UAU" | "UAC" -> "Tyrosine"
        | "UGU" | "UGC" -> "Cysteine"
        | "UGG" -> "Tryptophan"
        | "UAA" | "UAG" | "UGA" -> "STOP"
        | _ -> failwith $"unknown codon {code}"

let proteins rna = 
    Seq.chunkBySize 3 rna 
    |> Seq.map (System.String >> translate) 
    |> Seq.takeWhile (fun x -> x <> "STOP")
    |> Seq.toList